lib/winapi/filemanip.c

```diff
@@ -235,3 +236,106 @@ BOOL MoveFileA (LPCTSTR lpExistingFileName, LPCTSTR lpNewFileName)
+        return FALSE;
+    }
+
+    status = NtQueryVolumeInformationFile(handle, &ioStatusBlock, &fsSizeInfo, sizeof(fsSizeInfo), FileFsSizeInformation);
+    if (!NT_SUCCESS(status)) {
+        SetLastError(RtlNtStatusToDosError(status));
+        return FALSE;
+    }
+
+    status = NtClose(handle);
+    if (!NT_SUCCESS(status)) {
+        SetLastError(RtlNtStatusToDosError(status));
+        return FALSE;
+    }
+
+    ULONGLONG bytesPerAllocUnit = fsSizeInfo.BytesPerSector * fsSizeInfo.SectorsPerAllocationUnit;
+    ULONGLONG totalBytes = bytesPerAllocUnit * fsSizeInfo.TotalAllocationUnits.QuadPart;
+    ULONGLONG freeBytes = bytesPerAllocUnit * fsSizeInfo.AvailableAllocationUnits.QuadPart;
+
+    if (lpFreeBytesAvailableToCaller) {
+        lpFreeBytesAvailableToCaller->QuadPart = freeBytes;
+    }
+    if (lpTotalNumberOfBytes) {
+        lpTotalNumberOfBytes->QuadPart = totalBytes;
+    }
+    if (lpTotalNumberOfFreeBytes) {
+        lpTotalNumberOfFreeBytes->QuadPart = freeBytes;
+    }
+
+    return TRUE;
+}
+
+BOOL GetDiskFreeSpaceA (LPCSTR lpRootPathName, LPDWORD lpSectorsPerCluster, LPDWORD lpBytesPerSector, LPDWORD lpNumberOfFreeClusters, LPDWORD lpTotalNumberOfClusters)
+{
+    NTSTATUS status;
+    HANDLE handle;
+    ANSI_STRING path;
+    OBJECT_ATTRIBUTES objectAttributes;
+    IO_STATUS_BLOCK ioStatusBlock;
+    FILE_FS_SIZE_INFORMATION fsSizeInfo;
+
+    assert(lpRootPathName);
+
+    RtlInitAnsiString(&path, lpRootPathName);
+    InitializeObjectAttributes(&objectAttributes, &path, OBJ_CASE_INSENSITIVE, ObDosDevicesDirectory(), NULL);
+
+    status = NtOpenFile(&handle, FILE_LIST_DIRECTORY | SYNCHRONIZE, &objectAttributes, &ioStatusBlock, FILE_SHARE_READ | FILE_SHARE_WRITE, FILE_SYNCHRONOUS_IO_NONALERT | FILE_DIRECTORY_FILE | FILE_OPEN_FOR_FREE_SPACE_QUERY);
+    if (!NT_SUCCESS(status)) {
+        SetLastError(RtlNtStatusToDosError(status));
+        if (GetLastError() == ERROR_FILE_NOT_FOUND) {
+            SetLastError(ERROR_PATH_NOT_FOUND);
+        }
+        return FALSE;
+    }
+
+    status = NtQueryVolumeInformationFile(handle, &ioStatusBlock, &fsSizeInfo, sizeof(fsSizeInfo), FileFsSizeInformation);
+    if (!NT_SUCCESS(status)) {
+        SetLastError(RtlNtStatusToDosError(status));
+        return FALSE;
+    }
+
+    status = NtClose(handle);
+    if (!NT_SUCCESS(status)) {
+        SetLastError(RtlNtStatusToDosError(status));
+        return FALSE;
+    }
+
+    if (lpSectorsPerCluster) {
+        *lpSectorsPerCluster = fsSizeInfo.SectorsPerAllocationUnit;
+    }
+    if (lpBytesPerSector) {
+        *lpBytesPerSector = fsSizeInfo.BytesPerSector;
+    }
+    if (lpNumberOfFreeClusters) {
+        *lpNumberOfFreeClusters = fsSizeInfo.AvailableAllocationUnits.HighPart == 0 ? fsSizeInfo.AvailableAllocationUnits.LowPart : 0xFFFFFFFF;
+    }
+    if (lpTotalNumberOfClusters) {
+        *lpTotalNumberOfClusters = fsSizeInfo.TotalAllocationUnits.HighPart == 0 ? fsSizeInfo.TotalAllocationUnits.LowPart : 0xFFFFFFFF;
```

Comment:
nit: `fsSizeInfo.TotalAllocationUnits.HighPart ? 0xFFFFFFFF : fsSizeInfo.TotalAllocationUnits.LowPart;`

This helps with readability in my opinion (as the connection between highpart and maxing out the value is more apparent). It's also shorter and doesn't add more operations into the ternary operator (which users often struggle with anyway).
Also matches more closely with the implicit `!= NULL` in the assert for the pathname above.

*(Other instance, too)*

Reply:
Good point, I changed it.